pkg/p2p/server.go

```diff
@@ -0,0 +1,845 @@
+		zap.Int64("epoch", peer.Epoch))
+	m.peerLock.Lock()
+	delete(m.peers, peer.PeerID)
+	m.peerLock.Unlock()
+	if err != nil {
+		peer.abortWithError(ctx, err)
+	}
+}
+
+// MustAddHandler registers a handler for messages in a given topic and waits for the operation
+// to complete.
+func (m *MessageServer) MustAddHandler(
+	ctx context.Context,
+	topic string,
+	tpi interface{},
+	fn func(string, interface{}) error,
+) (<-chan error, error) {
+	doneCh, errCh, err := m.AddHandler(ctx, topic, tpi, fn)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	select {
+	case <-ctx.Done():
+		return nil, errors.Trace(ctx.Err())
+	case <-doneCh:
+	case <-m.closeCh:
+		return nil, cerror.ErrPeerMessageServerClosed.GenWithStackByArgs()
+	}
+	return errCh, nil
+}
+
+// AddHandler registers a handler for messages in a given topic.
+func (m *MessageServer) AddHandler(
+	ctx context.Context,
+	topic string,
+	tpi interface{},
+	fn func(string, interface{}) error) (chan struct{}, <-chan error, error) {
+	tp := reflect.TypeOf(tpi)
+
+	metricsServerRepeatedMessageCount := serverRepeatedMessageCount.MustCurryWith(prometheus.Labels{
+		"topic": topic,
+	})
+
+	poolHandle := m.pool.RegisterEvent(func(ctx context.Context, argsI interface{}) error {
+		args := argsI.(poolEventArgs)
+		sm := args.streamMeta
+		entry := args.entry
+		e := reflect.New(tp.Elem()).Interface()
+
+		m.acksMapLock.Lock()
+		lastAck := m.getAck(sm.SenderId, entry.GetTopic())
+		if lastAck >= entry.Sequence {
+			metricsServerRepeatedMessageCount.With(prometheus.Labels{
+				"from": sm.SenderAdvertisedAddr,
+			}).Inc()
+
+			log.Debug("skipping peer message",
+				zap.String("sender-id", sm.SenderId),
+				zap.String("topic", topic),
+				zap.Int64("skipped-Seq", entry.Sequence),
+				zap.Int64("last-ack", lastAck))
+			m.acksMapLock.Unlock()
+			return nil
+		}
+
+		if lastAck != 0 && entry.Sequence > lastAck+1 {
+			log.Panic("seq is skipped. Bug?", zap.Int64("last-ack", lastAck))
+		}
+
+		m.acksMapLock.Unlock()
+
+		if err := unmarshalMessage(entry.Content, e); err != nil {
+			return cerror.WrapError(cerror.ErrPeerMessageDecodeError, err)
+		}
+
+		if err := fn(sm.SenderId, e); err != nil {
+			return errors.Trace(err)
+		}
+
+		m.acksMapLock.Lock()
+		m.setAck(sm.SenderId, entry.GetTopic(), entry.GetSequence())
+		m.acksMapLock.Unlock()
+
+		return nil
+	}).OnExit(func(err error) {
+		log.Warn("topic handler returned error", zap.Error(err))
+		_ = m.scheduleTask(ctx, taskOnDeregisterHandler{
+			topic: topic,
+		})
+	})
+
+	doneCh := make(chan struct{})
+
+	if err := m.scheduleTask(ctx, taskOnRegisterHandler{
+		topic:   topic,
+		handler: poolHandle,
+		done:    doneCh,
+	}); err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	return doneCh, poolHandle.ErrCh(), nil
+}
+
+// MustRemoveHandler removes the registered handler for the given topic and wait
+// for the operation to complete.
+func (m *MessageServer) MustRemoveHandler(ctx context.Context, topic string) error {
```

Comment:
why we need two methods?

Reply:
`MustRemoveHandler` is implemented using `RemoveHandler`. The latter is an asynchronous version.